// react-native-blob-courier/lib/commonjs/Consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_PROGRESS_UPDATE_INTERVAL_MILLISECONDS = exports.DEFAULT_FILE_MULTIPART_FIELD_NAME = exports.DEFAULT_FETCH_TARGET = exports.BLOB_UPLOAD_FALLBACK_PARAMETERS = exports.BLOB_MULTIPART_UPLOAD_FALLBACK_PARAMETERS = exports.BLOB_FETCH_FALLBACK_PARAMETERS = exports.BLOB_COURIER_PROGRESS_EVENT_NAME = exports.ANDROID_DOWNLOAD_MANAGER_FALLBACK_PARAMETERS = void 0;

/**
 * Copyright (c) Ely Deckers.
 *
 * This source code is licensed under the MPL-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
const ANDROID_DOWNLOAD_MANAGER_FALLBACK_PARAMETERS = {
  description: undefined,
  enableNotifications: true,
  title: undefined
};
exports.ANDROID_DOWNLOAD_MANAGER_FALLBACK_PARAMETERS = ANDROID_DOWNLOAD_MANAGER_FALLBACK_PARAMETERS;
const DEFAULT_FETCH_TARGET = 'cache';
exports.DEFAULT_FETCH_TARGET = DEFAULT_FETCH_TARGET;
const DEFAULT_PROGRESS_UPDATE_INTERVAL_MILLISECONDS = 200;
exports.DEFAULT_PROGRESS_UPDATE_INTERVAL_MILLISECONDS = DEFAULT_PROGRESS_UPDATE_INTERVAL_MILLISECONDS;
const DEFAULT_FILE_MULTIPART_FIELD_NAME = 'file';
exports.DEFAULT_FILE_MULTIPART_FIELD_NAME = DEFAULT_FILE_MULTIPART_FIELD_NAME;
const BLOB_FETCH_FALLBACK_PARAMETERS = {
  android: {
    downloadManager: ANDROID_DOWNLOAD_MANAGER_FALLBACK_PARAMETERS,
    target: DEFAULT_FETCH_TARGET,
    useDownloadManager: false
  },
  headers: {},
  ios: {
    target: DEFAULT_FETCH_TARGET
  },
  method: 'GET',
  progressIntervalMilliseconds: DEFAULT_PROGRESS_UPDATE_INTERVAL_MILLISECONDS
};
exports.BLOB_FETCH_FALLBACK_PARAMETERS = BLOB_FETCH_FALLBACK_PARAMETERS;
const BLOB_UPLOAD_FALLBACK_PARAMETERS = {
  headers: {},
  method: 'POST',
  returnResponse: false
};
exports.BLOB_UPLOAD_FALLBACK_PARAMETERS = BLOB_UPLOAD_FALLBACK_PARAMETERS;
const BLOB_MULTIPART_UPLOAD_FALLBACK_PARAMETERS = { ...BLOB_UPLOAD_FALLBACK_PARAMETERS,
  parts: {}
};
exports.BLOB_MULTIPART_UPLOAD_FALLBACK_PARAMETERS = BLOB_MULTIPART_UPLOAD_FALLBACK_PARAMETERS;
const BLOB_COURIER_PROGRESS_EVENT_NAME = 'BlobCourierProgress';
exports.BLOB_COURIER_PROGRESS_EVENT_NAME = BLOB_COURIER_PROGRESS_EVENT_NAME;
